test/extensions/stats_sinks/hystrix/hystrix_test.cc

```diff
@@ -0,0 +1,377 @@
+    }
+  }
+
+  std::unordered_map<std::string, std::string> buildClusterMap(absl::string_view data_message) {
+    std::unordered_map<std::string, std::string> cluster_message_map;
+    std::vector<std::string> messages =
+        absl::StrSplit(data_message, "data:", absl::SkipWhitespace());
+    for (auto message : messages) {
+      if (absl::StrContains(getStreamField(message, "type"), "HystrixCommand")) {
+        std::string cluster_name(getStringStreamField(message, "name"));
+        cluster_message_map[cluster_name] = message;
+      }
+    }
+    return cluster_message_map;
+  }
+
+  TestRandomGenerator rand_;
+  uint64_t window_size_ = rand_.random() % 10 + 5; // Arbitrary reasonable number.
+  const std::string cluster1_name_{"test_cluster1"};
+  ClusterTestInfo cluster1_{cluster1_name_};
+
+  NiceMock<Http::MockStreamDecoderFilterCallbacks> callbacks_;
+  NiceMock<Server::MockInstance> server_;
+  Upstream::ClusterManager::ClusterInfoMap cluster_map_;
+
+  std::unique_ptr<HystrixSink> sink_;
+  NiceMock<Stats::MockSource> source_;
+  NiceMock<Upstream::MockClusterManager> cluster_manager_;
+};
+
+TEST_F(HystrixSinkTest, EmptyFlush) {
+  InSequence s;
+  Buffer::OwnedImpl buffer = createClusterAndCallbacks();
+  // Register callback to sink.
+  sink_->registerConnection(&callbacks_);
+  sink_->flush(source_);
+  std::string data_message = TestUtility::bufferToString(buffer);
+  validateResults(data_message, 0, 0, 0, 0, 0, window_size_);
+}
+
+TEST_F(HystrixSinkTest, BasicFlow) {
+  InSequence s;
+  Buffer::OwnedImpl buffer = createClusterAndCallbacks();
+  // Register callback to sink.
+  sink_->registerConnection(&callbacks_);
+
+  // Only success traffic, check randomly increasing traffic
+  // Later in the test we'll "shortcut" by constant traffic
+  uint64_t traffic_counter = 0;
+  ;
+  sink_->flush(source_); // init window with 0
+  for (uint64_t i = 0; i < (window_size_ - 1); i++) {
+    buffer.drain(buffer.length());
+    traffic_counter += rand_.random() % 1000;
+    ON_CALL(cluster1_.success_counter_, value()).WillByDefault(Return(traffic_counter));
+    sink_->flush(source_);
+  }
+
+  EXPECT_EQ(getStreamField(TestUtility::bufferToString(buffer), "rollingCountSuccess"),
+            std::to_string(traffic_counter));
+  EXPECT_EQ(getStreamField(TestUtility::bufferToString(buffer), "requestCount"),
+            std::to_string(traffic_counter));
+  EXPECT_EQ(getStreamField(TestUtility::bufferToString(buffer), "errorCount"), "0");
+  EXPECT_EQ(getStreamField(TestUtility::bufferToString(buffer), "errorPercentage"), "0");
+
+  // check mixed traffic
+  // Arbitrary values for testing. Make sure error > timeout.
+  const uint64_t success_step = 13;
+  const uint64_t error_4xx_step = 12;
+  const uint64_t error_4xx_retry_step = 11;
+  const uint64_t error_5xx_step = 10;
+  const uint64_t error_5xx_retry_step = 9;
+  const uint64_t timeout_step = 8;
+  const uint64_t timeout_retry_step = 7;
+  const uint64_t rejected_step = 6;
```

Comment:
Why are these in descending order?

Maybe some comments about what these steps mean.  Are they units of time?  Are they deltas between one another or relative to a start time?

Reply:
maybe I can use random() to make you believe it is truly unimportant random values representing traffic stats?

Reply:
I just want comments explaining what these numbers mean generally. And you can understand that if I see them explicitly in descending order it makes it look like that's important, so it's easy to get confused reading the code.

Reply:
Would it be ok to use random() here?